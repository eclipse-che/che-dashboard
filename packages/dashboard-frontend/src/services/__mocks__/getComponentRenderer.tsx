/*
 * Copyright (c) 2018-2021 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';
import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';

export default function <T extends Array<unknown>>(
  getComponent: (..._args: T) => React.ReactElement,
): {
  createSnapshot: (...args: T) => renderer.ReactTestRenderer;
  renderComponent: (...args: T) => {
    reRenderComponent: (..._args: T) => void;
  };
} {
  return {
    createSnapshot: (...args: T) => {
      return renderer.create(getComponent(...args));
    },
    renderComponent: (...args: T) => {
      const res = render(getComponent(...args));
      return {
        reRenderComponent: (..._args: T) => {
          res.rerender(getComponent(..._args));
        },
      };
    },
  };
}
