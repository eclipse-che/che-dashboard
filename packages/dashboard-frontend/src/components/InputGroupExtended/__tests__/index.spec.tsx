/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { ValidatedOptions } from '@patternfly/react-core';
import * as React from 'react';

import getComponentRenderer, { fireEvent, screen } from '@/services/__mocks__/getComponentRenderer';

import { InputGroupExtended, Props } from '..';

const { createSnapshot, renderComponent } = getComponentRenderer(getComponent);

const mockOnRemove = jest.fn();

describe('InputGroupExtended', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('snapshots', () => {
    test('readonly', () => {
      const snapshot = createSnapshot({
        isLoading: false,
        readonly: true,
        required: false,
        value: 'value',
        onRemove: mockOnRemove,
      });

      expect(snapshot.toJSON()).toMatchSnapshot();
    });

    test('required', () => {
      const snapshot = createSnapshot({
        isLoading: false,
        readonly: false,
        required: true,
        value: 'value',
        onRemove: mockOnRemove,
      });

      expect(snapshot.toJSON()).toMatchSnapshot();
    });
  });

  it('should handle submiting', () => {
    renderComponent({
      isLoading: false,
      readonly: false,
      required: false,
      value: 'value',
      validated: ValidatedOptions.success,
      onRemove: mockOnRemove,
    });

    const input = screen.queryByTestId('text-input');
    fireEvent.submit(input!);
  });

  it('should handle removing the entry', () => {
    renderComponent({
      isLoading: false,
      readonly: false,
      required: false,
      value: 'value',
      validated: ValidatedOptions.success,
      onRemove: mockOnRemove,
    });

    const removeButton = screen.queryByTestId('button-remove');
    expect(removeButton).not.toBeNull();
    expect(removeButton).toBeEnabled();

    removeButton!.click();

    expect(mockOnRemove).toHaveBeenCalled();
  });

  it('should disable the remove button if required', () => {
    renderComponent({
      isLoading: false,
      readonly: false,
      required: true,
      value: 'value',
      validated: ValidatedOptions.default,
      onRemove: mockOnRemove,
    });

    const removeButton = screen.queryByTestId('button-remove');
    expect(removeButton).not.toBeNull();
    expect(removeButton).toBeDisabled();
  });

  it('should disable the remove button if in progress', () => {
    renderComponent({
      isLoading: true,
      readonly: false,
      required: false,
      value: 'value',
      validated: ValidatedOptions.default,
      onRemove: mockOnRemove,
    });

    const removeButton = screen.queryByTestId('button-remove');
    expect(removeButton).not.toBeNull();
    expect(removeButton).toBeDisabled();
  });
});

function getComponent(props: Props): React.ReactElement {
  return (
    <InputGroupExtended {...props}>
      <input data-testid="text-input" />
    </InputGroupExtended>
  );
}
