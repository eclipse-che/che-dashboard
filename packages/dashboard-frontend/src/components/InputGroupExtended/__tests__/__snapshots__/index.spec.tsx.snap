// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputGroupExtended snapshots readonly 1`] = `
<span
  className="readonly"
>
  value
</span>
`;

exports[`InputGroupExtended snapshots required 1`] = `
<div
  className="pf-c-input-group nameInput"
>
  <input
    data-testid="text-input"
  />
  <button
    aria-disabled={true}
    aria-label={null}
    className="pf-c-button pf-m-plain pf-m-disabled"
    data-ouia-component-id="OUIA-Generated-Button-plain-1"
    data-ouia-component-type="PF4/Button"
    data-ouia-safe={true}
    data-testid="button-remove"
    disabled={true}
    onClick={[Function]}
    role={null}
    tabIndex={null}
    type="button"
  >
    <svg
      aria-hidden={true}
      aria-labelledby={null}
      fill="currentColor"
      height="1em"
      role="img"
      style={
        {
          "verticalAlign": "-0.125em",
        }
      }
      viewBox="0 0 512 512"
      width="1em"
    >
      <path
        d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zM124 296c-6.6 0-12-5.4-12-12v-56c0-6.6 5.4-12 12-12h264c6.6 0 12 5.4 12 12v56c0 6.6-5.4 12-12 12H124z"
      />
    </svg>
  </button>
</div>
`;
